src/variational/VariationalInference.jl

```diff
@@ -0,0 +1,171 @@
+module Variational
+
+using ..Core, ..Core.RandomVariables, ..Utilities
+using Distributions, Bijectors
+using ProgressMeter, LinearAlgebra
+using ..Turing: PROGRESS
+using ..Turing: Model, SampleFromPrior, SampleFromUniform
+using ..Turing: Turing
+using Random: AbstractRNG
+
+using ForwardDiff
+using Flux.Tracker, Flux.Optimise
+
+import ..Core: getchunksize, getADtype
+
+export
+    vi,
+    ADVI,
+    ELBO
+
+
+abstract type VariationalInference{AD} end
+
+getchunksize(::T) where {T <: VariationalInference} = getchunksize(T)
+getchunksize(::Type{<:VariationalInference{AD}}) where AD = getchunksize(AD)
+getADtype(alg::VariationalInference) = getADtype(typeof(alg))
+getADtype(::Type{<: VariationalInference{AD}}) where {AD} = AD
+
+abstract type VariationalObjective end
+
+const VariationalPosterior = Distribution{Multivariate, Continuous}
+
+"""
+    rand(vi::VariationalInference, num_samples)
+
+Produces `num_samples` samples for the given VI method using number of samples equal to `num_samples`.
+"""
+function rand(vi::VariationalPosterior, num_samples) end
```

Comment:
We probably want to throw an error as callback.

Reply:
This will just throw an error if there's nothing implemented though, right?

Reply:
Glad you brought up this because I deliberately didn't implement throwing an error for these as I'm not sure we should even have them.

`rand` should already be "enforced" by the fact that `VariationalPosterior <: Distribution`. I put it there before making `VariationalPosterior` an alias to `Distribution{Multivariate, Continuous}`. Should I just remove it for now? 

We can add it back if we move away from aliasing `Distribution`.